machine-learning/app/models/clip.py

```diff
@@ -78,55 +60,177 @@ def _predict(self, image_or_text: Image.Image | str) -> list[float]:
             case Image.Image():
                 if self.mode == "text":
                     raise TypeError("Cannot encode image as text-only model")
-                pixel_values = self.transform(image_or_text)
-                assert isinstance(pixel_values, torch.Tensor)
-                pixel_values = torch.unsqueeze(pixel_values, 0).numpy()
-                outputs = self.vision_model.run(self.vision_outputs, {"pixel_values": pixel_values})
+
+                outputs = self.encode_image(image_or_text)
```

Comment:
The abstract methods `tokenize` and `transform`are not used in the abstract base class.
I think it makes more sense to put the call to `self.transform(image_or_text)` here, and `self.tokenize(image_or_text)` a few lines below.
Also directly calling `self.vision_mode.run(...`. So the methods `encode_image` and `encode_text` are not needed at all.
This removes some redundancy when multiple implementations (in the future) would have the same `encode_x` methods.

Reply:
Oh, good point. `encode_image` and `encode_text` used to handle preprocessing too, but they're redundant now that `tokenize` and `transform` are separate methods.